pkg/sql/sem/asof/utils.go

```diff
@@ -0,0 +1,75 @@
+// Copyright 2023 The Cockroach Authors.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.txt.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0, included in the file
+// licenses/APL.txt.
+
+package asof
+
+import (
+	"context"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/eval"
+	"github.com/cockroachdb/cockroach/pkg/sql/sem/tree"
+	"github.com/cockroachdb/cockroach/pkg/sql/types"
+	"github.com/cockroachdb/cockroach/pkg/util/hlc"
+	"github.com/cockroachdb/errors"
+)
+
+// TypeCheckSystemTimeExpr type checks an Expr as a system time. It
+// accepts the same types as AS OF SYSTEM TIME expressions and
+// functions that evaluate to one of those types.
+//
+// The types need to be kept in sync with those supported by
+// asof.DatumToHLC.
+//
+// TODO(ssd): AOST and SPLIT are restricted to the use of constant expressions
+// or particular follower-read related functions. Do we want to do that here as well?
+// One nice side effect of allowing functions is that users can use NOW().
```

Comment:
why do we even need our own separate helpers? why can't we type check arbitrary timestamps like changefeeds do?

https://github.com/cockroachdb/cockroach/blob/50210bab0785fe871934a52ab4b1f5d0743c0709/pkg/ccl/changefeedccl/changefeed_stmt.go#L430

Reply:
🤷 The first option is that I just didn't think of doing that. However, if I reconstruct my reasoning my guess is that at the time I was experimenting seeing if I could get placeholders to work nicely and making sure we accepted the same set of types accepted by AsOf and then I didn't circle back to make it nicer.

If we do use asof.Eval, we'll want some new eval options to allow arbitrary functions (or at least NOW()) as well as an option to pass a different usage to DatumToHLC.